### ISSUE TITLE
Ability for GLTFDocumentExtension to add new nodes into middle of scene tree

### ISSUE THREAD

**Issue by eoineoineoin** [godotengine/godot-proposals#5891]
### Describe the project you are working on

Developing a plugin to import data from a glTF extension.

### Describe the problem or limitation you are having in your project

I'm attempting to use a GLTFDocumentExtension to add a Node3D to the _middle_ of a scene tree.

For example, I have a glTF file which, when imported has a scene tree which looks like like:

```
ObjectA : MeshInstance3D
    ObjectB: MeshInstance3D
        ObjectC: MeshInstance3D
```

I need to handle the case where `ObjectB` has extension which requires it to have a different parent, so I want to generate a scene with the layout:

```
ObjectA : MeshInstance3D
    GeneratedObject : MyCustomNode3D
        ObjectB: MeshInstance3D
            ObjectC: MeshInstance3D
```

Using GLTFDocumentExtension::_generate_scene_node(), I can currently write something like:

```
func _generate_scene_node(state : GLTFState, gltfNode : GLTFNode, sceneParent : Node) -> Node3D:
       if needToGenerateObject():
               var generated = MyCustomNode3D.new()
               generated.name = "GeneratedObject"
               .....
               return generated
```

However, this causes the "ObjectB : MeshInstance3D" to be dropped from the scene.

This is due to the following code in gltf_document.cpp (snipped for brevity):

```
void GLTFDocument::_generate_scene_node(Ref<GLTFState> state, Node *scene_parent, Node3D *scene_root, const GLTFNodeIndex node_index) {

    Node3D *current_node = nullptr;
    for (Ref<GLTFDocumentExtension> ext : document_extensions) {
           current_node = ext->generate_scene_node(state, gltf_node, scene_parent);
    }

    // If none of our GLTFDocumentExtension classes generated us a node, we generate one.
    if (!current_node) {
         current_node = _generate_mesh_instance(state, node_index);
    }
```

As far as I can tell, there's no way to generate the node which the GLTFDocument _would_ have made if the extension had not returned null, so I can't add that node as a child of my generated node.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Ideally, GLTFDocument::_generate_scene_node() would either:

1. Always generate the "default" node and pass it in to the GLTFDocumentExtension::generate_scene_node
2. Pass in some additional state or callback which could be used to generate that node

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I'd suggest that GLTFDocument::_generate_scene_node be modified to swap the order of the blocks beginning with the comments

```
// Check if any GLTFDocumentExtension classes want to generate a node for us.
```
and
```
// If none of our GLTFDocumentExtension classes generated us a node, we generate one.
```

then, subsequently, add an additional argument to pass `current_node` into GLTFDocumentExtension::generate_scene_node()

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Cannot be easily replicated by authors of GLTFDocumentExtension - even if so, would likely result in mistakes and duplicated code by anyone who implements such a similar extension.

### Is there a reason why this should be core and not an add-on in the asset library?

Functionality is _almost_ already available; this is a small, incremental improvement, which helps plugin authors.

**Comment by eoineoineoin** [comment]
On reflection, I think there's actually a bug here:

```
    for (Ref<GLTFDocumentExtension> ext : document_extensions) {
           current_node = ext->generate_scene_node(state, gltf_node, scene_parent);
    }
```

If you have _multiple_ GLTFDocumentExtension registered for a document, if any of them implement generate_scene_node(), then any generated node will be thrown away (except for the last-registered extension).

**Comment by fire** [comment]
@aaronfranke Since we were involved, I nudge you.

**Comment by aaronfranke** [comment]
> If you have multiple GLTFDocumentExtension registered for a document, if any of them implement generate_scene_node(), then any generated node will be thrown away (except for the last-registered extension).

No. It will break once a node is generated. You have this misconception because you're only looking at the snippet you posted, but the full code actually looks like this:

```cpp
	for (Ref<GLTFDocumentExtension> ext : document_extensions) {
		ERR_CONTINUE(ext.is_null());
		current_node = ext->generate_scene_node(state, gltf_node, scene_parent);
		if (current_node) {
			break;
		}
	}
```

> Ideally, GLTFDocument::_generate_scene_node() would either:
> 1. Always generate the "default" node and pass it in to the GLTFDocumentExtension::generate_scene_node

No, we don't want this. It would mean that we generate nodes which are likely to be unused.

What you're asking for opens up a can of worms if we tried to do anything "automatically". What if a GLTF file provides a mesh, camera, light, extension, all on the same GLTF node? Right now what would happen is we generate only the first match, so for example a GLTF node with both a mesh and a camera will never generate the camera. What *should* happen here? Does GLTF have any rules on what we are supposed to do in this case? Should we generate all nodes and add them as a child of each other? Which gets to be the parent then, or should there just be a blank parent?

Can you elaborate more of your use case? Why do you have an extension on a node which also has a mesh? As I mentioned above, if we want to support this fully, it goes beyond just generating one other node.

>   2. Pass in some additional state or callback which could be used to generate that node

Let me know if this will work for your use case: We can move the MeshInstance node generation logic into an exposed method, such as creating `GLTFMesh::to_node` similar to the existing methods on GLTFCamera and GLTFLight. Then, if an extension expects to handle data on the same node as a mesh, that extension can run `GLTFMesh::to_node` as needed and add that as a child. This would greatly simplify this can of worms, but note that it will result in a scene structure like this:

```
ObjectA : MeshInstance3D
    GeneratedObject : MyCustomNode3D // Generated in the extension, this is the returned node.
        ObjectB: MeshInstance3D // Generated in the extension together with GeneratedObject, added as a child.
        ObjectC: MeshInstance3D // Generated later by GLTFDocument.
```

**Comment by eoineoineoin** [comment]
> No. It will break once a node is generated. You have this misconception because you're only looking at the snippet you posted, but the full code actually looks like this:

You're right, I was not completely correct here, but that behaviour still feels a little weird to me? It means that some generate_scene_node() won't be called, so, depending on the order you register youre extensions, you'll get a different set of output functionality, and from the user's perspective, some extensions will appear to be broken.

> What if a GLTF file provides a mesh, camera, light, extension, all on the same GLTF node? Right now what would happen is we generate only the first match, so for example a GLTF node with both a mesh and a camera will never generate the camera. ..... Which gets to be the parent then, or should there just be a blank parent?

Yeah, this is kinda awkward, but what you're describing is fundamentally what I'm struggling with. I have a node which is a mesh+myObject, and there's no prohibition against such constructs in glTF. I think this is just a complication of Godot's inheritance-based scene representation - if I was using -say- Unity3D, a GameObject(==Node3D) can have a mesh component and a camera component simultaneously.

It's something I've struggled with when using Godot in my personal projects; I usually rely on enforcing a particular hierarchy and naming convention for child nodes. I'm not sure how people doing more "professional" projects tend do to things like this?

> Can you elaborate more of your use case? Why do you have an extension on a node which also has a mesh?

In this particular case, I'm trying to implement a glTF extension for rigid body physics (https://github.com/eoineoineoin/glTF_Physics) - in the Godot representation of a tree, the parent object is required to be a RigidBody3D, but I still need to have the child MeshInstance3D (or even a Node3D, with child MeshInstance3Ds, etc.) in order for the object to be displayed to the user. In this use-case, the parent-child relationship _is_ significant, due to the effect of RigidBody3D on the transform of the object, but I think that's not _necessarily_ the case for a general glTF extension, e.g. an audio source.

>  We can move the MeshInstance node generation logic into an exposed method, such as creating GLTFMesh::to_node similar to the existing methods on GLTFCamera and GLTFLight.

Yes, that's certainly an improvement and would make it possible to implement my extension. However, I'm still a little concerned about the "multiple extension" problem. With a solution like that, every extension author would write something like:

```
func _generate_scene_node(state : GLTFState, gltfNode : GLTFNode, sceneParent : Node) -> Node3D:
       if needToGenerateObject():
               var generated = MyCustomNode3D.new()
               generated.name = "GeneratedObject"
               generated.add_child( gltfNode.to_node() )
               .....
               return generated
```

Then, if multiple generate_scene_node() _could_ be called for multiple GLTFDocumentExtension, you'd end up with multiple instances of the original object.

To me, it really feels like the GLTFDocument loading code should read something like this pseudocode:

```
Node3D* current_node = make_node_or_camera_or_mesh()
for( GLTFDocumentExtension* extension : extensions )
      current_node = extension->generate_scene_node( current_node )
````

That way, at least all the generate_scene_node()s would be called - you'd still have the ordering problem, if you're dealing with extensions which have a significant effect in that regard; but at least that's configurable by the user and they'll get all the nodes for the extensions they expect.

**Comment by fire** [comment]
There seems to be a disagreement and the discussion has stalled.

**Comment by aaronfranke** [comment]
@fire While I disagree with some of the solutions proposed due to doing too much automatically, I think that there are totally improvements we can make to at least make it possible to support this use case. I will tackle this issue sometime this year, maybe in a few months, for the moment I have other priorities.

**Comment by aaronfranke** [comment]
I have a good idea of how to do solve this, but I haven't done it yet because I am waiting on smaller PRs to be merged first. However, my goal is still to solve this in 2023.

Basically, instead of generating nodes until we get one and then stopping, we need to change the logic a bit so that we keep generating nodes, and then every node after the first node is added as a child.

Example use case and behavior: For a GLTF file with a node structure like this:

```
A # Plain node
    B # Mesh and Physics Body
        C # Plain node
```

We would generate a node structure like this:

```
A # Plain node
    B # Physics Body
        BMesh # Mesh
        C # Plain node
```

This preserves the GLTF node hierarchy exactly, just with some extra leaf nodes added. So unlike what's proposed in the OP, a reference to a node path of `^"A/B/C"` will not be affected if you add another "component" to B. Also, this does not break compatibility vs the current master because in master it will generate this:

```
A # Plain node
    B # Physics Body only, the mesh is discarded :(
        C # Plain node
```

The way this works is that we first check all extensions, then we generate stuff from the base GLTF spec afterwards. Therefore the first extension that generates a node has that node "be" the node in the tree. We will ensure that physics is always run first, except the user is allowed to register a custom extension as high priority, in which case it would run before all built-in node generation including physics. That is already the case in master and in the proposed behavior. The only difference with the proposed behavior is that once a node is generated, don't stop generating.
